Fix triton server doc (#52665)
ray-project/ray: doc/source/serve/tutorials/triton-server-integration.md

This guide shows how to build an application with stable diffusion model using [NVIDIA Triton Server](https://github.com/triton-inference-server/server) in Ray Serve.

## Preparation

> Serving a model with Triton Inference Server isn't as straightforward. Compared to vanilla Ray Serve, it requires jumping through quite a few hoops:

Reviewer: I'd soften the language, I think people will understand based on the steps :)  
```suggestion
NVIDIA Triton provides a python API for integration with Ray Serve. This can allow users to leverage inference optimizations available in Triton with the development experience and multi-model capabilities of Ray Serve. These are the steps required to prepare your model for using Triton with Ray Serve, with the TensorRT backend:
1. Set up a local or remote model repository.
2. Convert your model into the ONNX format.
3. After exporting to ONNX, convert it into a TensorRT engine serialized file.
4. Setup inference with Ray Serve.

Alternatively, you can also look at other inference optimizations like `torch.compile` to use with Ray Serve.
```

    output_names=["image"],
    do_constant_folding=True,
    export_params=True,
)

logger.info("VAE decoder exported successfully")
logger.info("Export complete. Models saved to model_repository/stable_diffusion/1/")
```
From the script, the outputs are `text_encoder.onnx`, `unet.onnx` `vae_decoder.onnx`.

After the ONNX model exported, convert the ONNX model to the TensorRT engine serialized file. ([Details](https://github.com/NVIDIA/TensorRT/blob/release/9.2/samples/trtexec/README.md?plain=1#L22) about trtexec cli)

```bash
# Convert text encoder
trtexec --onnx=model_repository/stable_diffusion/1/text_encoder.onnx \
        --saveEngine=model_repository/stable_diffusion/1/text_encoder.engine \
        --fp16 \
        --workspace=4096

# Convert UNet
trtexec --onnx=model_repository/stable_diffusion/1/unet.onnx \
        --saveEngine=model_repository/stable_diffusion/1/unet.engine \
        --fp16 \
        --workspace=4096

# Convert VAE decoder
trtexec --onnx=model_repository/stable_diffusion/1/vae_decoder.onnx \
        --saveEngine=model_repository/stable_diffusion/1/vae_decoder.engine \
        --fp16 \
        --workspace=4096
```

## Part 2: Serving

> It is recommended to use the `nvcr.io/nvidia/tritonserver:23.12-py3` image which already has the Triton Server python API library installed, and install the ray serve lib by `pip install "ray[serve]"` inside the image.

Reviewer: ```suggestion
It is recommended to use the `nvcr.io/nvidia/tritonserver:23.12-py3` image (or later versions) which has the Triton Server python API library installed, and install the ray serve lib by `pip install "ray[serve]"` inside the image.
```

This guide shows how to build an application with stable diffusion model using [NVIDIA Triton Server](https://github.com/triton-inference-server/server) in Ray Serve.

## Preparation
NVIDIA Triton provides a python API for integration with Ray Serve. This can allow users to leverage inference optimizations available in Triton with the development experience and multi-model capabilities of Ray Serve. These are the steps required to prepare your model for using Triton with Ray Serve, with the TensorRT backend:
1. Set up a local or remote model repository.
2. Convert your model into the ONNX format.
3. After exporting to ONNX, convert it into a TensorRT engine serialized file.
4. Setup inference with Ray Serve with a custom image

Alternatively, you can also consider and compare other inference optimizations techniques like `torch.compile` to use with Ray Serve.

> 1. Set up a local or remote model repository (it's not optional, details buried deeper in the tutorial).

Reviewer: Looks like these didn't get removed

Author: removed

2. Manually convert your model into the ONNX format, regardless of what framework you originally used.
3. After exporting to ONNX, convert it into a TensorRT engine serialized file.
4. Finally, once everything is properly formatted and in place, you can load the model for serving.

TL;DR, there's a lot of extra setup and preprocessing before you get to the actual serving part.

> ## Part I: Convert and serialize the model

Reviewer 2: Can we have a part 0 description the environment setup? I feel it's hard to understand if the user should start from a ray image or trition image or if there are some required dependencies that need to be installed ahead of time...

# Export text encoder
torch.onnx.export(
    text_encoder,
    dummy_text_input,
    str(model_path / 'text_encoder.onnx'),
    opset_version=14,
    input_names=["input_ids"],
    output_names=["last_hidden_state"],
    do_constant_folding=True,
    export_params=True,
)

logger.info("Text encoder exported successfully")

# Export UNet
torch.onnx.export(
    unet,
    (dummy_sample, timestep, encoder_hidden_states),
    str(model_path / 'unet.onnx'),
    opset_version=14,
    input_names=["sample", "timestep", "encoder_hidden_states"],
    output_names=["out_sample"],
    do_constant_folding=True,
    export_params=True,
)

logger.info("UNet exported successfully")

# Export VAE decoder
dummy_latent = torch.randn(batch_size, 4, 64, 64, device=device)

torch.onnx.export(
    pipe.vae.decoder,
    dummy_latent,
    str(model_path / 'vae_decoder.onnx'),
    opset_version=14,
    input_names=["latent"],
    output_names=["image"],
    do_constant_folding=True,
    export_params=True,
)

logger.info("VAE decoder exported successfully")
logger.info("Export complete. Models saved to model_repository/stable_diffusion/1/")
```

> From the script, the outputs are `text_encoder.onnx`, `unet.onnx` `vae_decoder.onnx`.

Reviewer 3: ```suggestion
From the script, the outputs are `text_encoder.onnx`, `unet.onnx`, and `vae_decoder.onnx`.
```

> After the ONNX model exported, convert the ONNX model to the TensorRT engine serialized file. ([Details](https://github.com/NVIDIA/TensorRT/blob/release/9.2/samples/trtexec/README.md?plain=1#L22) about trtexec cli)

Reviewer 3: ```suggestion
After the ONNX model is exported, convert the ONNX model to the TensorRT engine serialized file. ([Details](https://github.com/NVIDIA/TensorRT/blob/release/9.2/samples/trtexec/README.md?plain=1#L22) about trtexec cli)
```

Reviewer 2: Would be more readable if we link to the markdown
```suggestion
After the ONNX model exported, convert the ONNX model to the TensorRT engine serialized file. ([Details](https://github.com/NVIDIA/TensorRT/blob/release/9.2/samples/trtexec/README.md#description) about trtexec cli)
```

Also I'm sure you have tested it working with 9.2, but was this intentional to link to an older version where the current latest is 10.9?

```bash
# Convert text encoder
trtexec --onnx=model_repository/stable_diffusion/1/text_encoder.onnx \
        --saveEngine=model_repository/stable_diffusion/1/text_encoder.engine \
        --fp16 \
        --workspace=4096

# Convert UNet
trtexec --onnx=model_repository/stable_diffusion/1/unet.onnx \
        --saveEngine=model_repository/stable_diffusion/1/unet.engine \
        --fp16 \
        --workspace=4096

# Convert VAE decoder
trtexec --onnx=model_repository/stable_diffusion/1/vae_decoder.onnx \
        --saveEngine=model_repository/stable_diffusion/1/vae_decoder.engine \
        --fp16 \
        --workspace=4096
```

## Part 2: Serving

> It is recommended to use the `nvcr.io/nvidia/tritonserver:23.12-py3` image (or later versions) which has the Triton Server python API library installed, and install the ray serve lib by `pip install "ray[serve]"` inside the image.

Reviewer 3: ```suggestion
It is recommended to use the `nvcr.io/nvidia/tritonserver:23.12-py3` image (or later versions) which has the Triton Server python API library installed, and install Ray Serve by running `pip install "ray[serve]"` inside the image.
```

Reviewer 2: I feel we should not suggest later version if it's not tested
```suggestion
It is recommended to use the `nvcr.io/nvidia/tritonserver:23.12-py3` image which has the Triton Server python API library installed, and install the ray serve lib by `pip install "ray[serve]"` inside the image.
```

Also, noticed this is the same version as when the doc is first created. I'm sure you have tested it working, but is the latest image (25.04-py3) not working?

```python
import torch

from pathlib import Path
from diffusers import StableDiffusionPipeline

# Load a specific model version that's known to work well with ONNX conversion

> model_id = "runwayml/stable-diffusion-v1-5"  # This is often the most compatible

Reviewer 2: Just double check that you have this model tested exist? if not maybe we can use this one https://huggingface.co/stable-diffusion-v1-5/stable-diffusion-v1-5/tree/main

Author: I have a workspace i tested on. Is there any standard approach to attach workspace examples to PR description in such cases?

Reviewer 2: Usually I just take screenshots. But don't worry I believe you, just also see https://github.com/huggingface/optimum-habana/issues/1305 and was wondering how this work XD
